File: Maximum Depth of Binary Tree LC-104.py
# TC => O(N) SC => O(H), H is the max height/depth of the tree

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, root, n):
      if not root:
        return n
      return max(self.dfs(root.left, n+1),self.dfs(root.right, n+1))
    def maxDepth(self, root: TreeNode) -> int:
        return self.dfs(root, 0)

File: Maximum Depth of Binary Tree LC-104.cpp
// tc=o(n)
// sc=o(h)
class Solution {
public:
    int solve(TreeNode* root, int lvl){
        if(!root) return lvl;
        return max(solve(root->left,lvl+1),solve(root->right, lvl+1));
    }
    int maxDepth(TreeNode* root) {
        return solve(root,0);
    }
};